### DISCUSSION TITLE
Unable to launch container docker

### DISCUSSION THREAD

**Discussion by tottidici** [ThinkR-open/golem#760] [Q&A]
I created an app with golem. I want to use docker but the container won't launch. There is an error :
`Error in loadNamespace(name) : there is no package called ‘demogolem’
Calls: :: ... loadNamespace -> withRestarts -> withOneRestart -> doWithOneRestart
Execution halted`

here is the dockerfile : 
`
FROM rocker/r-ver:3.6.3
RUN apt-get update && apt-get install -y git-core libcurl4-openssl-dev libgit2-dev libglpk-dev libgmp-dev libicu-dev libssl-dev libxml2-dev make pandoc pandoc-citeproc zlib1g-dev && rm -rf /var/lib/apt/lists/*
RUN echo "options(repos = c(CRAN = 'https://cran.rstudio.com/'), download.file.method = 'libcurl', Ncpus = 4)" >> /usr/local/lib/R/etc/Rprofile.site
RUN R -e 'install.packages("remotes")'
RUN Rscript -e 'remotes::install_version("glue",upgrade="never", version = "1.4.2")'
RUN Rscript -e 'remotes::install_version("processx",upgrade="never", version = "3.5.1")'
RUN Rscript -e 'remotes::install_version("htmltools",upgrade="never", version = "0.5.1.1")'
RUN Rscript -e 'remotes::install_version("dplyr",upgrade="never", version = "1.0.5")'
RUN Rscript -e 'remotes::install_version("pkgload",upgrade="never", version = "1.2.1")'
RUN Rscript -e 'remotes::install_version("attempt",upgrade="never", version = "0.3.1")'
RUN Rscript -e 'remotes::install_version("shiny",upgrade="never", version = "1.6.0")'
RUN Rscript -e 'remotes::install_version("DBI",upgrade="never", version = "1.1.1")'
RUN Rscript -e 'remotes::install_version("testthat",upgrade="never", version = "3.0.2")'
RUN Rscript -e 'remotes::install_version("config",upgrade="never", version = "0.3.1")'
RUN Rscript -e 'remotes::install_version("spelling",upgrade="never", version = "2.2")'
RUN Rscript -e 'remotes::install_version("shinydashboard",upgrade="never", version = "0.7.1")'
RUN Rscript -e 'remotes::install_version("shinycssloaders",upgrade="never", version = "1.0.0")'
RUN Rscript -e 'remotes::install_version("RPostgreSQL",upgrade="never", version = "0.6-2")'
RUN Rscript -e 'remotes::install_version("highcharter",upgrade="never", version = "0.8.2")'
RUN Rscript -e 'remotes::install_version("golem",upgrade="never", version = "0.3.1")'
RUN Rscript -e 'remotes::install_version("DT",upgrade="never", version = "0.18")'
RUN mkdir /build_zone
ADD . /build_zone
WORKDIR /build_zone
RUN R -e 'remotes::install_local(upgrade="never")'
RUN rm -rf /build_zone
EXPOSE 80
CMD R -e "options('shiny.port'=80,shiny.host='0.0.0.0');demogolem::run_app()"
`
Thanks for your help

**Comment by VincentGuyader** [comment]
can you try with this one :  


```
FROM rocker/r-ver:3.6.3
RUN apt-get update && apt-get install -y git-core libcurl4-openssl-dev libgit2-dev libglpk-dev libgmp-dev libicu-dev libssl-dev libxml2-dev make pandoc pandoc-citeproc zlib1g-dev && rm -rf /var/lib/apt/lists/*
RUN echo "options(repos = c(CRAN = 'https://packagemanager.rstudio.com/all/__linux__/focal/latest'), download.file.method = 'libcurl', Ncpus = 4)" >> /usr/local/lib/R/etc/Rprofile.site
RUN echo "options(warn = 2);" >> $R_HOME/etc/Rprofile.site
RUN R -e 'install.packages("remotes")'
RUN Rscript -e 'remotes::install_version("glue",upgrade="never", version = "1.4.2")'
RUN Rscript -e 'remotes::install_version("processx",upgrade="never", version = "3.5.1")'
RUN Rscript -e 'remotes::install_version("htmltools",upgrade="never", version = "0.5.1.1")'
RUN Rscript -e 'remotes::install_version("dplyr",upgrade="never", version = "1.0.5")'
RUN Rscript -e 'remotes::install_version("pkgload",upgrade="never", version = "1.2.1")'
RUN Rscript -e 'remotes::install_version("attempt",upgrade="never", version = "0.3.1")'
RUN Rscript -e 'remotes::install_version("shiny",upgrade="never", version = "1.6.0")'
RUN Rscript -e 'remotes::install_version("DBI",upgrade="never", version = "1.1.1")'
RUN Rscript -e 'remotes::install_version("testthat",upgrade="never", version = "3.0.2")'
RUN Rscript -e 'remotes::install_version("config",upgrade="never", version = "0.3.1")'
RUN Rscript -e 'remotes::install_version("spelling",upgrade="never", version = "2.2")'
RUN Rscript -e 'remotes::install_version("shinydashboard",upgrade="never", version = "0.7.1")'
RUN Rscript -e 'remotes::install_version("shinycssloaders",upgrade="never", version = "1.0.0")'
RUN Rscript -e 'remotes::install_version("RPostgreSQL",upgrade="never", version = "0.6-2")'
RUN Rscript -e 'remotes::install_version("highcharter",upgrade="never", version = "0.8.2")'
RUN Rscript -e 'remotes::install_version("golem",upgrade="never", version = "0.3.1")'
RUN Rscript -e 'remotes::install_version("DT",upgrade="never", version = "0.18")'
RUN mkdir /build_zone
ADD . /build_zone
WORKDIR /build_zone
RUN R -e 'remotes::install_local(upgrade="never")'
RUN rm -rf /build_zone
EXPOSE 80
CMD R -e "options('shiny.port'=80,shiny.host='0.0.0.0');demogolem::run_app()"
```


and tell me wich step fails ?

**Comment by tottidici** [comment]
Hi

I'm trying your dockerfile
I took the opportunity to upgrade to R 4.0.5 !! and now there is an error at the RPostgreSQL install stage

`
 => ERROR [19/27] RUN Rscript -e 'remotes::install_version("RPostgreSQL",upgrade="never", version = "0.7-  7.4s
------
 > [19/27] RUN Rscript -e 'remotes::install_version("RPostgreSQL",upgrade="never", version = "0.7-2")':
#22 2.408 Downloading package from url: https://packagemanager.rstudio.com/all/__linux__/focal/latest/src/contrib/RPostgreSQL_0.7-2.tar.gz
#22 6.298 Installing package into ‘/usr/local/lib/R/site-library’
#22 6.298 (as ‘lib’ is unspecified)
#22 6.544 * installing *source* package ‘RPostgreSQL’ ...
#22 6.552 ** package ‘RPostgreSQL’ successfully unpacked and MD5 sums checked
#22 6.553 ** using staged installation
#22 6.660 checking for gcc... gcc
#22 6.679 checking whether the C compiler works... yes
#22 6.708 checking for C compiler default output file name... a.out
#22 6.710 checking for suffix of executables...
#22 6.740 checking whether we are cross compiling... no
#22 6.775 checking for suffix of object files... o
#22 6.795 checking whether we are using the GNU C compiler... yes
#22 6.811 checking whether gcc accepts -g... yes
#22 6.829 checking for gcc option to accept ISO C89... none needed
#22 6.866 checking build system type... x86_64-pc-linux-gnu
#22 6.907 checking host system type... x86_64-pc-linux-gnu
#22 6.908 checking target system type... x86_64-pc-linux-gnu
#22 6.908 checking for pg_config... no
#22 6.909 configure: checking for PostgreSQL header files
#22 6.909 configure: Checking include /usr/include.
#22 6.909 configure: Checking include /usr/include/postgresql.
#22 6.909 configure: Checking include /usr/local/include.
#22 6.909 configure: Checking include /usr/local/include/postgresql.
#22 6.909 configure: Checking include /usr/local/pgsql/include.
#22 6.909 configure: Checking include /usr/local/postgresql/include.
#22 6.910 configure: Checking lib /usr/lib.
#22 6.910 configure: Checking lib /usr/lib/postgresql.
#22 6.910 configure: Checking lib /usr/local/lib.
#22 6.910 configure: Checking lib /usr/local/lib/postgresql.
#22 6.910 configure: Checking lib /usr/local/postgresql/lib.
#22 6.916 conftest.c:1:10: fatal error: libpq-fe.h: No such file or directory
#22 6.916     1 | #include <libpq-fe.h>
#22 6.916       |          ^~~~~~~~~~~~
#22 6.916 compilation terminated.
#22 6.952 checking for "/libpq-fe.h"... no
#22 6.966 configure: creating ./config.status
#22 7.036 config.status: creating src/Makevars
#22 7.074 ** libs
#22 7.079 gcc -I"/usr/local/lib/R/include" -DNDEBUG -I  -I/usr/local/include   -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c RS-DBI.c -o RS-DBI.o
#22 7.370 gcc -I"/usr/local/lib/R/include" -DNDEBUG -I  -I/usr/local/include   -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c RS-PQescape.c -o RS-PQescape.o
#22 7.377 In file included from RS-PQescape.c:7:
#22 7.377 RS-PostgreSQL.h:23:14: fatal error: libpq-fe.h: No such file or directory
#22 7.377    23 | #    include "libpq-fe.h"
#22 7.377       |              ^~~~~~~~~~~~
#22 7.377 compilation terminated.
#22 7.379 make: *** [/usr/local/lib/R/etc/Makeconf:172: RS-PQescape.o] Error 1
#22 7.380 ERROR: compilation failed for package ‘RPostgreSQL’
#22 7.380 * removing ‘/usr/local/lib/R/site-library/RPostgreSQL’
#22 7.391 Error: Failed to install 'unknown package' from URL:
#22 7.391   (converted from warning) installation of package ‘/tmp/RtmpYnu2yJ/remotes778284941/RPostgreSQL’ had non-zero exit status
#22 7.391 Execution halted
------
executor failed running [/bin/sh -c Rscript -e 'remotes::install_version("RPostgreSQL",upgrade="never", version = "0.7-2")']: exit code: 1
`

**Reply by tottidici** [reply]
this is dockerfile
`FROM rocker/r-ver:4.0.5
RUN apt-get update && apt-get install -y git-core libcurl4-openssl-dev libgit2-dev libglpk-dev libgmp-dev libicu-dev libssl-dev libxml2-dev make pandoc pandoc-citeproc zlib1g-dev && rm -rf /var/lib/apt/lists/*
RUN echo "options(repos = c(CRAN = 'https://packagemanager.rstudio.com/all/__linux__/focal/latest'), download.file.method = 'libcurl', Ncpus = 4)" >> /usr/local/lib/R/etc/Rprofile.site
RUN echo "options(warn = 2);" >> $R_HOME/etc/Rprofile.site
RUN R -e 'install.packages("remotes")'
RUN Rscript -e 'remotes::install_version("glue",upgrade="never", version = "1.4.2")'
RUN Rscript -e 'remotes::install_version("processx",upgrade="never", version = "3.5.2")'
RUN Rscript -e 'remotes::install_version("htmltools",upgrade="never", version = "0.5.2")'
RUN Rscript -e 'remotes::install_version("dplyr",upgrade="never", version = "1.0.7")'
RUN Rscript -e 'remotes::install_version("pkgload",upgrade="never", version = "1.2.2")'
RUN Rscript -e 'remotes::install_version("attempt",upgrade="never", version = "0.3.1")'
RUN Rscript -e 'remotes::install_version("shiny",upgrade="never", version = "1.7.1")'
RUN Rscript -e 'remotes::install_version("DBI",upgrade="never", version = "1.1.1")'
RUN Rscript -e 'remotes::install_version("testthat",upgrade="never", version = "3.1.0")'
RUN Rscript -e 'remotes::install_version("config",upgrade="never", version = "0.3.1")'
RUN Rscript -e 'remotes::install_version("spelling",upgrade="never", version = "2.2")'
RUN Rscript -e 'remotes::install_version("shinydashboard",upgrade="never", version = "0.7.2")'
RUN Rscript -e 'remotes::install_version("shinycssloaders",upgrade="never", version = "1.0.0")'
RUN Rscript -e 'remotes::install_version("RPostgreSQL",upgrade="never", version = "0.7-2")'
RUN Rscript -e 'remotes::install_version("highcharter",upgrade="never", version = "0.8.2")'
RUN Rscript -e 'remotes::install_version("golem",upgrade="never", version = "0.3.1")'
RUN Rscript -e 'remotes::install_version("DT",upgrade="never", version = "0.19")'
RUN mkdir /build_zone
ADD . /build_zone
WORKDIR /build_zone
RUN R -e 'remotes::install_local(upgrade="never")'
RUN rm -rf /build_zone
EXPOSE 80
CMD R -e "options('shiny.port'=80,shiny.host='0.0.0.0');demogolem::run_app()"
`

**Comment by tottidici** [comment]
it's ok!!

j'ai ajouté ces 2 éléments : `libsodium-dev libssl-dev`

now my dockerfile  looks like this

```
FROM rocker/r-ver:4.0.5
RUN apt-get update && apt-get install -y git-core libcurl4-openssl-dev libgit2-dev libglpk-dev libgmp-dev libicu-dev libpq-dev libsodium-dev libssl-dev libxml2-dev make pandoc pandoc-citeproc zlib1g-dev && rm -rf /var/lib/apt/lists/*
RUN echo "options(repos = c(CRAN = 'https://packagemanager.rstudio.com/all/__linux__/focal/latest'), download.file.method = 'libcurl', Ncpus = 4)" >> /usr/local/lib/R/etc/Rprofile.site
RUN echo "options(warn = 2);" >> $R_HOME/etc/Rprofile.site
RUN R -e 'install.packages("remotes")'
RUN Rscript -e 'remotes::install_version("glue",upgrade="never", version = "1.4.2")'
RUN Rscript -e 'remotes::install_version("processx",upgrade="never", version = "3.5.2")'
RUN Rscript -e 'remotes::install_version("htmltools",upgrade="never", version = "0.5.2")'
RUN Rscript -e 'remotes::install_version("dplyr",upgrade="never", version = "1.0.7")'
RUN Rscript -e 'remotes::install_version("pkgload",upgrade="never", version = "1.2.2")'
RUN Rscript -e 'remotes::install_version("attempt",upgrade="never", version = "0.3.1")'
RUN Rscript -e 'remotes::install_version("shiny",upgrade="never", version = "1.7.1")'
RUN Rscript -e 'remotes::install_version("DBI",upgrade="never", version = "1.1.1")'
RUN Rscript -e 'remotes::install_version("testthat",upgrade="never", version = "3.1.0")'
RUN Rscript -e 'remotes::install_version("config",upgrade="never", version = "0.3.1")'
RUN Rscript -e 'remotes::install_version("spelling",upgrade="never", version = "2.2")'
RUN Rscript -e 'remotes::install_version("shinydashboard",upgrade="never", version = "0.7.2")'
RUN Rscript -e 'remotes::install_version("shinycssloaders",upgrade="never", version = "1.0.0")'
RUN Rscript -e 'remotes::install_version("RPostgreSQL", upgrade="never", version = "0.7-2")'
RUN Rscript -e 'remotes::install_version("highcharter",upgrade="never", version = "0.8.2")'
RUN Rscript -e 'remotes::install_version("golem",upgrade="never", version = "0.3.1")'
RUN Rscript -e 'remotes::install_version("DT",upgrade="never", version = "0.19")'
RUN mkdir /build_zone
ADD . /build_zone
WORKDIR /build_zone
RUN R -e 'remotes::install_local(upgrade="never")'
RUN rm -rf /build_zone
EXPOSE 80
CMD R -e "options('shiny.port'=80,shiny.host='0.0.0.0');demogolem::run_app()"
```
